### ISSUE TITLE
codestyle issues in android code

### ISSUE THREAD

**Issue by pgrandin** [navit-gps/navit#832]
https://www.codefactor.io/repository/github/navit-gps/navit/pull/827

**Comment by jandegr** [comment]
or you can rely on these instead when it concerns java
https://15831-30791823-gh.circle-artifacts.com/0/reports/checkstyleMain.html
https://15831-30791823-gh.circle-artifacts.com/0/reports/lint-results.html

codefactor reports 71 java issues when there are actually 327, but codefactor seems to find new issues from time to time that were there already years ago and lint gives more android specific issues.

**Comment by jandegr** [comment]
I will post an update soon in https://github.com/navit-gps/navit/pull/863 that brings the codefactor issues in android down to 2, at least for the java part of the code,  brings the checkstyle issues mentioned above down from 327 to zero and a modest reduction in the lint warnings from 108 down to 101. 68 of the remaing lint warnings relate to https://github.com/navit-gps/navit/issues/867

Does the title of this issue imply that for C anything goes ? 
seems to be confirmed by several merges into trunk that create new codefactor issues, in added code, not just some mishaps in the way codefactor reports.

I am very eager to keep my own code clean but I ask the management to bring clarity because I am not going to port those cleanups over to Navit as a form of occupational therapy.

A remark for the non-believers, solving those issues actually leads to better working, better readable and better maintainable code. As a small sample, an invalid name often made me notice that it is just the acces modifier that is too wide. Similar for remarks about missing doc, often leads to a method that can be private. 

And finally a note for the do-believers : if you work on a chunck of code that causes a cyclomatic complexity issue, as soon as you change just one char, codefactor marks the old issue as resolved and reports it as a new issue caused by your commit.

thx.

**Comment by jandegr** [comment]
Another note for the do-believers (in the biblical sense) , others please hit the `back` or `home` button of your browser right now.

Since I could fix all but one of the android related codestyle issues, there is time and space to elaborate a bit.

Recently I stumbled onto a bit of code in a PR reported by codefactor with a complexity of 17.
To my best knowledge this is only 1 above the reporting threshold, lowhanging fruit to fix it.
Visual inspection and running a tool on it both reveal a cyclomatic complexity of 15. I am still figuring out what else codefactor might add to it or in what mode they scan it.

It is however normal for services to report the highest number if there are options, they must do that to be on the safe side, but you can be sure if it reports a chunk of code, then there is room for improvement.
 
By looking at the code a little longer I could reduce cyclomatic complexity with 4, so down to 11.
Codefactor then does not report it anymore but unfortunatly does not tel me how much it reduces overall complexity, but the -3 fix alone is already enough for codefactor. Since I'm interested in code quality and not in just running some scan service, I added the other fix too.

I abstracted the code a little bit and included a before and an after version. 
[refactor.pdf](https://github.com/navit-gps/navit/files/3597284/refactor.pdf)

I was already surprised that a seasoned dev got frustrated over it and left a nasty remark  about it, but that somebody gives a thumbsup for such remarks really puzzles me.

**Comment by metalstrolch** [comment]
Now for the "non believers" or so.
First of all, I like to play with open cards. The "seasoned dev" is me. The corresponding pull request is  #870.

I like static code analysis a lot, and tools like CodeFactor may report many usefull findings and usually it's a good thing to honour that.
But cyclomatic complexity as a metric is the biggest junk since using lines of code as metric for productivity. One example out of the attached PDF file shows best what I mean:
Converting a simple sanity check:
```
if((r == NULL) || (in == NULL) || (out == NULL) || (count_out == NULL) || (*count_out < count_in*8+1)) {
    return;
}
```
into
```
int has_nullpointer = (r == NULL) + (in == NULL) + (out == NULL) + (count_out == NULL);
if(has_nullpointer || (*count_out < count_in*8+1)) {
    return;
}
```
just to reduce the count of || and &&. I don't understand in what way this code is less error prone, better to test or even just better to read. Just to reduce cyclomatic complexity.
Even McCabe, the inventor of the metric wrote in his essay (on page 314):

> The particular upper bound that has been used for cyclomatic complexity is 10 which seems like a reasonable, but not magical,upper limit [1]

So he himself was not sure what a good upper limit was. And btw, he did only count branches, not !! and &&.

About #870: This is a fix based on existing code for a quite annoying problem. I tried to solve this being as fast as possible while requiring as less as possible stack and ram. I was happy seeing it running after some thinkering. And on the box doing this, in my precious free time for "fun", I don't have a toolkit counting cyclomatic complexity. And even worse, every tool counts different.
Now, while on the branch, CodeFactor doesn't even run. Then, finally happy having solved the problem, one makes a pull request learning that CodeFactor thinks this is too complex. In fact, CodeFactor doesn't even tell you what the limit is. I don't fix such problems as "occupational therapy" as well, given the fact that this helps a back end (Windows CE) I'm not at all personally interested in. And I don't refactor the already working code that way, creating artificial silly statements, just to reduce the branch and !! or && count. BTW: the old implementation for polygons (without holes) had a memory leak. Nobody, and no tool found it before. I'm leaving that as an exercise for the readers to find it. 

About the earlier mentioned #861 and #862: Here I wrote code for quite specific graphics back ends for which I do not even own devices to test. The changes were based on existing obviously working code that was long inside the code base. I preferred to change that code as less as possible. and no, I won't screw it just because CodeFactor thinks it has a too high cyclomatic complexity. Still now I'm wondering that this code works without me even testing it a single time "live". 

I was thinking to somehow install myself an Android emulator and learn JNI to enable Android for multipolygons to fix #852 as well, but rantings like this here confirm me not to do. I have better things to do like writing code for closed architectures, and later being blamed for not obeying that silly cyclomatic complexity. I would't even know how to calculate it for object oriented languages like Java even worse for JNI. Maybe a exception counts twice...

We really should care for all those casts changing alignment rules, bad function pointers and so on prior wasting time and introducing errors doing refactorings like these proposed here. If somebody's dignity hangs on the CodeFactor result, he might happily kick in and fix my code. This is open source at least. Isn't it?

End of ranting
 - Metalstrolch

[1]: http://www.literateprogramming.com/mccabe.pdf
